File: Combinations.java
public class Solution {
    public ArrayList<ArrayList<Integer>> combine(int n, int k) {
        // Start typing your Java solution below
        // DO NOT write main() function
        ArrayList<Integer> combination = new ArrayList<Integer>();
        ArrayList<ArrayList<Integer>> ret = new ArrayList<ArrayList<Integer>>();
        if(k <= n)
        {
            buildCombine(1, n, k, combination, ret);
        }
        return ret;
    }
    
    private void buildCombine(int start, int end, int k, ArrayList<Integer> combination, ArrayList<ArrayList<Integer>> ret)
    {
        if(k==0)
        {
            ret.add((ArrayList<Integer>)combination.clone());
            return;
        }
        for(int i = start; i <= end; i++)
        {
            combination.add(i);
            buildCombine(i+1, end, k-1, combination, ret);
            int lastIndex = combination.size() - 1;
            combination.remove(lastIndex);
        }
    }
}

File: Combinations.cpp
class Solution {
public:
    vector<int> oneResult;
    vector<vector<int>>results;
    vector<vector<int> > combine(int n, int k) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function
        oneResult.clear();
        results.clear();
        buildCombination(1, n, k);
        return results;
    }
    
    
    void buildCombination(int index, int n, int k)
    {
        if(k==0)
        {
            results.push_back(oneResult);
            return;
        }
        
        for(int i = index; i <= n; i++)
        {
            oneResult.push_back(i);
            buildCombination(i+1, n, k-1);
            oneResult.pop_back();
        }
    }
};
